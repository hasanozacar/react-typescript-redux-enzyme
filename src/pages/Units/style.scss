.container {
  padding: 20px;
  display: flex;
  align-items: center;
}
.table {
  margin-top: 24px;
}

.cost-filter-element {
  display: flex;
  align-items: center;
}

.slider {
  width: 100%;
}
.buttonGroup {
  width: 100%;
}
.button {
  width: 20%;
}
.margin-horizontal15 {
  margin-left: 15px;
  margin-right: 15px;
}
