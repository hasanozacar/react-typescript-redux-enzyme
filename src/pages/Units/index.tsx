import React, { Component } from 'react';
import { connect } from "react-redux";
import { getUnitsAction } from '../../actions/units/get';
import { filterUnitAction } from '../../actions/units/filter';
import { Ages, Cost, FilterType, RootObject } from '../../types/dataTypes';
import { History } from 'history';
import { setUnitDetail } from '../../actions/units/set-unit-detail';
import { Button, Table, Checkbox, Slider, Avatar, Card } from 'antd';
import { withRouter } from 'react-router-dom';
import './style.scss';
import { AnyAction } from 'redux'
import sultan from '../../global/assets/sultan.jpeg'

const ButtonGroup = Button.Group;
export type UnitsState = {
  ages: Array<Ages>,
  filter: FilterType,
  filterDrawerVisible: boolean,
};

export type UnitsProps = {
  history: History,
  units: Array<RootObject>,
  getUnitsAction: (payload?: AnyAction | undefined) => void;
  filterUnitAction: (payload: FilterType) => void;
  setUnitDetail: (record: any) => void;
};
class Units extends Component<UnitsProps, UnitsState>  {
  constructor(props: UnitsProps) {
    super(props)

    this.state = {
      filterDrawerVisible: false,
      ages: [
        { id: 0, name: 'All' },
        { id: 1, name: 'Dark' },
        { id: 2, name: 'Feudal' },
        { id: 3, name: 'Castle' },
        { id: 4, name: 'Imperial' },],
      filter: {
        age: "All",
        woodValue: null,
        foodValue: null,
        goldValue: null,
        woodChecked: false,
        foodChecked: false,
        goldChecked: false,
      }
    };
  }

  componentDidMount() {
    this.props.getUnitsAction()
  }

  handleSetFilter = (key: any, value: any) => {
    this.setState(prevState => ({ filter: { ...prevState.filter, [key]: value } }), () => {
      this.props.filterUnitAction(this.state.filter);
    });
  }

  renderAge = (age: Ages) => (
    <Button
      className="button"
      key={age.id}
      type={age.name === this.state.filter.age ? "primary" : "ghost"}
      onClick={() => this.handleSetFilter("age", age.name)}>
      {age.name}
    </Button>
  );

  renderAgesFilter = (ages: Array<Ages>) => (
    <>
      <h3 className="ages-title"> Ages </h3>
      <ButtonGroup className="buttonGroup"  >
        {ages.map(age => this.renderAge(age))}
      </ButtonGroup>
    </>
  )

  renderCostsFilter = (name: string, key: any, sliderValue: number, checkKey: any) => (
    <div className="cost-filter-element">
      <Checkbox onChange={() => this.handleSetFilter(checkKey, !this.state.filter[checkKey])} />
      <div className="margin-horizontal15"> {name} </div>
      <div className="slider">
        <Slider
          onChange={(value) => this.handleSetFilter(key, value)}
          min={0}
          max={200}
          disabled={!this.state.filter[checkKey]}
        />
      </div>
      <div>{sliderValue}</div>
    </div>
  )

  renderCosts = (costs: Cost) => {
    let value = '';
    if (costs?.Wood) {
      value = `Wood: ${costs.Wood}`;
    } if (costs?.Food) {
      value = value?.length > 0 ? (`${value}, Food: ${costs.Food}`) : `Food: ${costs.Food}`;
    } if (costs?.Gold) {
      value = value?.length > 0 ? (`${value}, Gold: ${costs.Gold}`) : `Gold: ${costs.Gold}`;
    }
    return value ? value : 'Unknown';
  }

  columns() {
    return [
      {
        title: "Avatar",
        key: 'avatar',
        render: () => (
          <Avatar src={sultan} />
        ),
      },
      {
        title: "Id",
        key: 'id',
        render: ({ id }: any) => id
      },
      {
        title: "Name",
        key: 'name',
        render: ({ name }: any) => name
      },
      {
        title: "Age",
        key: 'age',
        render: ({ age }: any) => age

      },
      {
        title: "Costs",
        key: 'costs',
        render: ({ cost }: any) => this.renderCosts(cost)
      },
      {
        title: 'Action',
        dataIndex: '',
        key: 'x',
        render: (text: string, record: object) => (
          <Button color="#1890ff" id="increment" onClick={
            () => {
              this.props.setUnitDetail(record);
              this.props.history.push('/unit_detail', { detail: record });
            }
          }>
            Show Details
          </Button >
        ),
      }
    ];
  }

  renderTable = (data: Array<RootObject>) => (
    <div className="table">
      <Table
        size="small"
        pagination={{ pageSize: 20 }}
        // rowKey={unit => unit.id.toString()}
        columns={this.columns()}
        dataSource={data}
      />
    </div>
  )

  render() {
    const { ages } = this.state;
    const { units } = this.props;
    return (
      <div className="container">
        <div className="col-12" style={{ maxWidth: 720, width: "100%" }}>
          <Card>
            {this.renderAgesFilter(ages)}
            {this.renderCostsFilter("Wood", "woodValue", this.state.filter.cost?.wood, "woodChecked")}
            {this.renderCostsFilter("Food", "foodValue", this.state.filter.cost?.food, "foodChecked")}
            {this.renderCostsFilter("Gold", "goldValue", this.state.filter.cost?.gold, "goldChecked")}
          </Card>

          {this.renderTable(units)}
        </div>
      </div>

    );
  }
}

const mapDispatchToProps = {
  getUnitsAction,
  filterUnitAction,
  setUnitDetail,
}

const mapStateToProps = (state: any) => ({
  units: state.units?.filteredData
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Units))