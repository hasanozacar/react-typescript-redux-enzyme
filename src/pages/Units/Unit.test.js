import React from "react";
import Units from ".";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import "../../setupTests";
import * as Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { BrowserRouter as Router } from "react-router-dom";
import { Button, Checkbox, Slider, Table } from "antd";

Enzyme.configure({ adapter: new Adapter() });

describe("Units", () => {
  const mockStore = configureStore();
  const store = mockStore({
    news: {
      title: "Mock Title",
      urlToImage: "www.google.com",
      description: "Mock Description",
      url: "bbc.com",
    },
    loading: false,
  });
  const wrapper = mount(
    <Provider store={store}>
      <Router>
        <Units
          units={[
            {
              id: 104,
              name: "Elite Woad Raider",
              description: "Upgraded Woad Raider",
              expansion: "Age of Kings",
              age: "Imperial",
              cost: {
                Food: 65,
                Gold: 25,
              },
              build_time: 10,
              reload_time: 2,
              movement_rate: 1.38,
              line_of_sight: 5,
              hit_points: 80,
              attack: 13,
              armor: "0/1",
              attack_bonus: ["+3 eagles", "+3 buildings"],
            },
            {
              id: 104,
              name: "Elite Woad Raider",
              description: "Upgraded Woad Raider",
              expansion: "Age of Kings",
              age: "Imperial",
              cost: {
                Gold: 25,
              },
              build_time: 10,
              reload_time: 2,
              movement_rate: 1.38,
              line_of_sight: 5,
              hit_points: 80,
              attack: 13,
              armor: "0/1",
              attack_bonus: ["+3 eagles", "+3 buildings"],
            },
          ]}
        />
      </Router>
    </Provider>
  );
  it("renders without crashing", () => {
    expect(wrapper.find(Units).length).toEqual(1);
  });
  it("should click button render correctly table", () => {
    wrapper.find(Button).at(0).props().onClick();
    expect(wrapper.find(Table).length).toEqual(1);
  });
  it("should on change checkbox render correctly table", () => {
    wrapper.find(Checkbox).at(1).props().onChange();
    expect(wrapper.find(Table).length).toEqual(1);
  });
  it("should on change render correctly table", () => {
    wrapper.find(Slider).at(1).props().onChange();
    expect(wrapper.find(Table).length).toEqual(1);
  });
});
