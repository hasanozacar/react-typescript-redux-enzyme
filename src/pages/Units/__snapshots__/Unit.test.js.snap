// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Units should render correctly all button 1`] = `ReactWrapper {}`;

exports[`should render correctly <UnitDetail> 1`] = `ReactWrapper {}`;
